### ISSUE TITLE
Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.

### ISSUE THREAD

**Issue by isthaison** [callstack/react-native-paper#639]
### Your Environment

| software              | version
| --------------------- | -------
| ios or android        |
| react-native          |
| react-native-paper    |
| node                  |
| npm or yarn           |

**Comment by Trancever** [comment]
Please provide more informations.

**Comment by isthaison** [comment]
when i upgrade sdk 31 then facing problem

**Comment by satya164** [comment]
You can ignore the warnings. They won't affect the functionality.

**Comment by RajenderDandyal** [comment]
getting the same warning .. multiple times when app starts ... 
any clue ..How to remove this warning 
using Native-base custom theme + RNv0.57 +RNNv2

**Comment by satya164** [comment]
@RajenderDandyal It's not a bug. You can ignore it: https://github.com/facebook/metro/issues/287

**Comment by ethanke** [comment]
3 times warning each startup of my app. Even if it's not a functional problem, would be great to find a way to fix this warning, is not clean...

<img width="848" alt="Capture d’écran 2019-08-15 à 11 05 59" src="https://user-images.githubusercontent.com/14801923/63085142-1eb3d600-bf4d-11e9-9435-74a1e2d2fa42.png">

**Comment by Trancever** [comment]
@ethanke This will hide warnings `console.ignoredYellowBox = ['Require cycle: node_modules/react-native-paper'];`. We do not prioritize fixing this, because it's not really an issue.

**Comment by firofame** [comment]
Still getting this issue.

**Comment by sajaddp** [comment]
Ignoring warnings is not the right thing to do |:

**Comment by Entx** [comment]
Tengo un problema alguien podria ayudarme 
> Task :app:mergeDexDebug FAILED

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.0.1/userguide/command_line_interface.html#sec:command_line_warnings
368 actionable tasks: 21 executed, 347 up-to-date
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
D8: Cannot fit requested classes in a single dex file (# methods: 79766 > 65536 ; # fields: 90083 > 65536)
com.android.builder.dexing.DexArchiveMergerException: Error while merging dex archives: 
The number of method references in a .dex file cannot exceed 64K.
Learn how to resolve this issue at https://developer.android.com/tools/building/multidex.html
	at com.android.builder.dexing.D8DexArchiveMerger.getExceptionToRethrow(D8DexArchiveMerger.java:131)
	at com.android.builder.dexing.D8DexArchiveMerger.mergeDexArchives(D8DexArchiveMerger.java:118)
	at com.android.build.gradle.internal.transforms.DexMergerTransformCallable.call(DexMergerTransformCallable.java:102)
	at com.android.build.gradle.internal.tasks.DexMergingTaskRunnable.run(DexMergingTask.kt:444)
	at com.android.build.gradle.internal.tasks.Workers$ActionFacade.run(Workers.kt:335)
	at org.gradle.workers.internal.AdapterWorkAction.execute(AdapterWorkAction.java:50)
	at org.gradle.workers.internal.DefaultWorkerServer.execute(DefaultWorkerServer.java:50)
	at org.gradle.workers.internal.NoIsolationWorkerFactory$1$1$1.create(NoIsolationWorkerFactory.java:65)
	at org.gradle.workers.internal.NoIsolationWorkerFactory$1$1$1.create(NoIsolationWorkerFactory.java:61)
	at org.gradle.internal.classloader.ClassLoaderUtils.executeInClassloader(ClassLoaderUtils.java:98)
	at org.gradle.workers.internal.NoIsolationWorkerFactory$1$1.execute(NoIsolationWorkerFactory.java:61)
	at org.gradle.workers.internal.AbstractWorker$1.call(AbstractWorker.java:44)
	at org.gradle.workers.internal.AbstractWorker$1.call(AbstractWorker.java:41)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:416)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:406)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:102)
	at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
	at org.gradle.workers.internal.AbstractWorker.executeWrappedInBuildOperation(AbstractWorker.java:41)
	at org.gradle.workers.internal.NoIsolationWorkerFactory$1.execute(NoIsolationWorkerFactory.java:53)
	at org.gradle.workers.internal.DefaultWorkerExecutor$3.call(DefaultWorkerExecutor.java:217)
	at org.gradle.workers.internal.DefaultWorkerExecutor$3.call(DefaultWorkerExecutor.java:212)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at org.gradle.internal.work.DefaultConditionalExecutionQueue$ExecutionRunner.runExecution(DefaultConditionalExecutionQueue.java:215)
	at org.gradle.internal.work.DefaultConditionalExecutionQueue$ExecutionRunner.runBatch(DefaultConditionalExecutionQueue.java:164)
	at org.gradle.internal.work.DefaultConditionalExecutionQueue$ExecutionRunner.run(DefaultConditionalExecutionQueue.java:131)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.android.tools.r8.CompilationFailedException: Compilation failed to complete
	at com.android.tools.r8.utils.t.a(:55)
	at com.android.tools.r8.D8.run(:11)
	at com.android.builder.dexing.D8DexArchiveMerger.mergeDexArchives(D8DexArchiveMerger.java:116)
	... 34 more
Caused by: com.android.tools.r8.utils.AbortException: Error: null, Cannot fit requested classes in a single dex file (# methods: 79766 > 65536 ; # fields: 90083 > 65536)
	at com.android.tools.r8.utils.Reporter.a(:21)
	at com.android.tools.r8.utils.Reporter.a(:7)
	at com.android.tools.r8.dex.VirtualFile.a(:33)
	at com.android.tools.r8.dex.VirtualFile$h.a(:5)
	at com.android.tools.r8.dex.ApplicationWriter.a(:13)
	at com.android.tools.r8.dex.ApplicationWriter.write(:35)
	at com.android.tools.r8.D8.d(:44)
	at com.android.tools.r8.D8.b(:1)
	at com.android.tools.r8.utils.t.a(:23)
	... 36 more


FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:mergeDexDebug'.
> A failure occurred while executing com.android.build.gradle.internal.tasks.Workers$ActionFacade
   > com.android.builder.dexing.DexArchiveMergerException: Error while merging dex archives: 
     The number of method references in a .dex file cannot exceed 64K.
     Learn how to resolve this issue at https://developer.android.com/tools/building/multidex.html

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

**Comment by azuxx** [comment]
![Screenshot 2020-04-17 at 14 59 36](https://user-images.githubusercontent.com/16132913/79571819-1f909f00-80bc-11ea-8e81-eeccf340eb7b.png)

Still having the issue with RN 0.62.2

**Comment by Trancever** [comment]
We are not planning to fix this anytime soon. You should silence this particular warning with: `console.ignoredYellowBox = ['Require cycle: node_modules/react-native-paper'];`

**Comment by catalincic** [comment]
HI @Trancever, funny you just replied to this, was looking for a solution myself too to hide the warnings!

Do you mind letting me know where exactly to add `console.ignoredYellowBox = ['Require cycle: node_modules/react-native-paper'];` ?

I've also tried adding YellowBox.ignoreWarnings(...) to the top of my root index.js file.

Thanks, much appreciated!

**Comment by HunteRoi** [comment]
> HI @Trancever, funny you just replied to this, was looking for a solution myself too to hide the warnings!
> 
> Do you mind letting me know where exactly to add `console.ignoredYellowBox = ['Require cycle: node_modules/react-native-paper'];` ?
> 
> I've also tried adding YellowBox.ignoreWarnings(...) to the top of my root index.js file, but it's not working.
> 
> Thanks, much appreciated!

@catalincic Same here! Mention me whenever you get an answer please :eyes:

**Comment by kulikalov** [comment]
> @ethanke This will hide warnings `console.ignoredYellowBox = ['Require cycle: node_modules/react-native-paper'];`. We do not prioritize fixing this, because it's not really an issue.

what do you mean it's not an issue? Quite a few people on the internet are spending their time on this. Fix the import, is it such a big deal to not fix it for 2 years??

**Comment by pke** [comment]
This is not an issue to be ignored. It renders the app in the simulator totally unresponsive! One can't even remove the in app warnings with "dismiss all".

**Comment by patheticGeek** [comment]
> This is not an issue to be ignored. It renders the app in the simulator totally unresponsive! One can't even remove the in app warnings with "dismiss all".

Put
`console.ignoredYellowBox = ['Require cycle: node_modules/react-native-paper'];`
in the entry file eg. index.js It removes these warnings from popping up in the yellow blox

**Comment by pke** [comment]
LOL.  No way! That is an ugly hack. 
The source cause needs to be fixed.

-----Original Message-----

**Comment by raajnadar** [comment]
@pke I fixed the issue in this PR https://github.com/callstack/react-native-paper/pull/1851 but it is not yet merged because the PR breaks the docs if you can help me we can get that PR merged.

**Comment by pke** [comment]
I'll take a look. 

-----Original Message-----

**Comment by cjduncana** [comment]
Que hubo @Entx 👋 Quizás estás buscando agregar `multiDexEnabled true` en tu `build.gradle` como descrito en [este artículo](https://developer.android.com/studio/build/multidex). Justo ayer me encontré con el mismo problema.

**Comment by amerllica** [comment]
This issue still exists  😢

**Comment by tondevelop** [comment]
I started to use RN 0.63.0. Still same warning exists. Maybe this warning comes from 'react-native-paper' I think.
Any solution will be appreciate. Thank you.
![image](https://user-images.githubusercontent.com/67708844/87494918-c0becf80-c682-11ea-9d4b-dfad6ad4a5ff.png)

**Comment by GanongLS** [comment]
using react native 0.63.0 still has the issue... 

![Screenshot from 2020-07-19 12-54-04](https://user-images.githubusercontent.com/39286664/87868338-f33c3580-c9be-11ea-8dad-5441a2d771a2.png)

**Comment by aavhanchatse** [comment]
Same issue. Please help.

**Comment by satsuya0114** [comment]
same issue. both in RN 0.61.5 & 0.63.0

**Comment by hkaracaoglu** [comment]
Same issue in RN 0.62.2

**Comment by GuilhermeCdrs** [comment]
Está aparecendo pra mim também, porem estou usando "react-native-maps".. Ainda não descobri a origem do erro/aviso.
![react-native-maps warn](https://user-images.githubusercontent.com/43662481/88470260-bac1bd80-ced0-11ea-9d58-426a649e765e.png)

**Comment by cjduncana** [comment]
> porem

Ninguém sabe.

**Comment by muhammadwafa** [comment]
Same issue in react-native 0.63.2

**Comment by CarlosNeto2804** [comment]
@GuilhermeCdrs Tente definir o estilo do MapView para uma posição absoluta com os valores superior, esquerdo, direito e inferior definidos.
Verifique se você ativou a API do Google Maps no console do desenvolvedor do Google;

**Comment by CarlosNeto2804** [comment]
const styles = StyleSheet.create({
  map: {
    ...StyleSheet.absoluteFillObject,
  },
});
<MapView
  style={styles.map}
  // other props
/>

**Comment by CarlosNeto2804** [comment]
> Same issue in react-native 0.63.2

@muhammadwafa  Try setting the style of the MapView to an absolute position with top, left, right and bottom values set.
Make sure you have enabled Google Maps API in Google developer console
const styles = StyleSheet.create({
  map: {
    ...StyleSheet.absoluteFillObject,
  },
});
<MapView
  style={styles.map}
  // other props
/>

**Comment by muhammadwafa** [comment]
I solved like this https://stackoverflow.com/a/63186739/11566074

**Comment by muhammadwafa** [comment]
I solved like this https://stackoverflow.com/a/63186739/11566074

**Comment by varnguyen** [comment]
I using react-native 0.63.2 still has this issue.

**Comment by varnguyen** [comment]
I think it is an issue print warning!!!

You are probably importing something from "file A" into "file B", then importing something again from "file B" into "file A". Examine all the imports from both the files and see if there's any such cycle.

**Comment by machine-dev** [comment]
@varnguyen Everybody knows that.
But this error is coming from `react-native-paper` and the PR is not merged yet.

**Comment by raulicious** [comment]
Since it's not fixed yet, is there a way to ignore these warnings in RN 0.63 that uses logBox instead of yellowBox?
```
LogBox.ignoreLogs([
    'Require cycle:'
]);
```
doesn't work for me

**Comment by ixez** [comment]
What the heck! It has been two years and this problem has not been solved yet!

**Comment by amerllica** [comment]
@ixez, And this is an Open-Source world.

**Comment by anonewb** [comment]
`LogBox.ignoreLogs(["Require cycle:"]);`

**Comment by ixez** [comment]
> `LogBox.ignoreLogs(["Require cycle:"]);`

It doesn't work for me. I found even I put `LogBox.ignoreAllLogs();`, the warning still comes out.

**Comment by ribeirogab** [comment]
The community is already two years old with this error and you do not solve it, regrettable...

**Comment by jaaywags** [comment]
I believe this has been fixed in this pull request, #2378 and released in [v4.5.0](https://github.com/callstack/react-native-paper/releases/tag/v4.5.0).

[Referencing this comment](https://github.com/callstack/react-native-paper/issues/2233#issuecomment-740694652).

**Comment by simonedavico** [comment]
@jaaywags it has not, 4.5.0 still produces the same warnings.

**Comment by mitchldtn** [comment]
I started getting this error after I added react-native-keychain, @apollo/client and graphql packages to my project.

I've been using '@react-navigation/material-bottom-tabs' prior to installing those packages without any of these errors.
If I stop using '@react-navigation/material-bottom-tabs' the errors disappear. 

I've been searching for a fix for this and all I'm seeing is to ignore the issue. Is this still the official response? If that is the case than my solution will be to stop using '@react-navigation/material-bottom-tabs' which requires the use of this package.

```
[Mon Jan 11 2021 19:48:16.669]  WARN     Require cycle: node_modules/react-native-paper/src/components/Checkbox/Checkbox.tsx -> node_modules/react-native-paper/src/components/Checkbox/CheckboxElements.tsx -> node_modules/react-native-paper/src/components/Checkbox/Checkbox.tsx

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
[Mon Jan 11 2021 19:48:16.674]  WARN     Require cycle: node_modules/react-native-paper/src/components/Checkbox/CheckboxElements.tsx -> node_modules/react-native-paper/src/components/Checkbox/CheckboxItem.tsx -> node_modules/react-native-paper/src/components/Checkbox/CheckboxElements.tsx

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
[Mon Jan 11 2021 19:48:16.675]  WARN     Require cycle: node_modules/react-native-paper/src/components/FAB/FABElements.tsx -> node_modules/react-native-paper/src/components/FAB/FABGroup.tsx -> node_modules/react-native-paper/src/components/FAB/FABElements.tsx

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
[Mon Jan 11 2021 19:48:16.679]  WARN     Require cycle: node_modules/react-native-paper/src/components/FAB/FAB.tsx -> node_modules/react-native-paper/src/components/FAB/FABElements.tsx -> node_modules/react-native-paper/src/components/FAB/FAB.tsx

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
[Mon Jan 11 2021 19:48:16.680]  WARN     Require cycle: node_modules/react-native-paper/src/components/RadioButton/RadioButton.tsx -> node_modules/react-native-paper/src/components/RadioButton/RadioButtonElements.tsx -> node_modules/react-native-paper/src/components/RadioButton/RadioButton.tsx

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
[Mon Jan 11 2021 19:48:16.681]  WARN     Require cycle: node_modules/react-native-paper/src/components/RadioButton/RadioButtonElements.tsx -> node_modules/react-native-paper/src/components/RadioButton/RadioButtonItem.tsx -> node_modules/react-native-paper/src/components/RadioButton/RadioButtonElements.tsx

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
[Mon Jan 11 2021 19:48:16.681]  WARN     Require cycle: node_modules/react-native-paper/src/components/Appbar/Appbar.tsx -> node_modules/react-native-paper/src/components/Appbar/AppbarElements.tsx -> node_modules/react-native-paper/src/components/Appbar/Appbar.tsx

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
[Mon Jan 11 2021 19:48:16.693]  WARN     Require cycle: node_modules/react-native-paper/src/components/Appbar/AppbarElements.tsx -> node_modules/react-native-paper/src/components/Appbar/AppbarBackAction.tsx -> node_modules/react-native-paper/src/components/Appbar/AppbarElements.tsx

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
[Mon Jan 11 2021 19:48:16.693]  WARN     Require cycle: node_modules/react-native-paper/src/components/Appbar/AppbarElements.tsx -> node_modules/react-native-paper/src/components/Appbar/AppbarHeader.tsx -> node_modules/react-native-paper/src/components/Appbar/AppbarElements.tsx

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
[Mon Jan 11 2021 19:48:16.694]  WARN     Require cycle: node_modules/react-native-paper/src/components/Appbar/Appbar.tsx -> node_modules/react-native-paper/src/components/Appbar/AppbarElements.tsx -> node_modules/react-native-paper/src/components/Appbar/AppbarHeader.tsx -> node_modules/react-native-paper/src/components/Appbar/Appbar.tsx

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
[Mon Jan 11 2021 19:48:16.694]  WARN     Require cycle: node_modules/react-native-paper/src/components/ToggleButton/ToggleButton.tsx -> node_modules/react-native-paper/src/components/ToggleButton/ToggleButtonElements.tsx -> node_modules/react-native-paper/src/components/ToggleButton/ToggleButton.tsx

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
[Mon Jan 11 2021 19:48:16.695]  WARN     Require cycle: node_modules/react-native-paper/src/components/ToggleButton/ToggleButtonElements.tsx -> node_modules/react-native-paper/src/components/ToggleButton/ToggleButtonRow.tsx -> node_modules/react-native-paper/src/components/ToggleButton/ToggleButtonElements.tsx

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
(node:10459) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 change listeners added to [HasteMap]. Use emitter.setMaxListeners() to increase limit
(Use `node --trace-warnings ...` to show where the warning was created)
```

**Comment by jaaywags** [comment]
As far as I can tell, this was actually fixed in Release [v4.7.0](https://github.com/callstack/react-native-paper/releases/tag/v4.7.0). I installed [v4.7.1](https://github.com/callstack/react-native-paper/releases/tag/v4.7.1) and no more require cycle messages!

Though, I did run into a weird issue. I got this error after upgrading

> unable to resolve module checkboxelements

I fixed it by clearing cache using this command
`npx react-native start --reset-cache`

Shoutout to [this post](https://github.com/callstack/react-native-paper/issues/1879#issuecomment-631995773).
